### PR TITLE
PuppiFatJets

### PR DESCRIPTION
**Pull request by sidnarayanan**
In parallel to #158 , NeroPuppiFatJets and BarePuppiFatJets classes have been implemented. Also implemented sequence to fill Puppi fat jets from Bambu.

### CODE REVIEW

**Comment by amarini** [comment]
From the miniaod part, we need to add the configuration line in Nero_cfy.py, and to append the analysis object in the Nero.cc plugin (otherwise they will not be executed).

**Comment by amarini** [comment]
Check compilation errors:
https://amarini.web.cern.ch/amarini/NeroProducer/f133f85b79e16da1388195c68c118b0c79d89ad2/build.txt

**Comment by sidnarayanan** [comment]
Sorry, forgot to commit and push this fix.

**Comment by sidnarayanan** [comment]
Looks like the 'CMS-standard' (i.e. Phil's) way of clustering Puppi jets from mini AOD is to re-run Puppi on top of slimmed PF candidates and feed this to a jet producer [1]. I started implementing this earlier today and will try to finish it tonight. I think ultimately we can improve on this method as discussed earlier, but this seems safe for now.

[1] https://github.com/ksung25/BaconProd/blob/master/Ntupler/config/makingBacon_MC_25ns_MINIAOD.py#L98-L100

**Comment by sidnarayanan** [comment]
@yiiyama the latest commit fixes a bug in which the jet energy correction object would be copied over when a MetCorrectionMod is cloned. This can cause things like the puppi MET to be doubly corrected (once for CHS jets and again for puppi jets). The solution is to do the cloning before creating the JEC object.

**Comment by yiiyama** [comment]
Green light for Bambu

**Comment by yiiyama** [comment]
What's the status with this PR?

File: Nero/test/testNero.py [hunk 1]
```diff
@@ -144,6 +144,7 @@
 ## SKIM INFO
 process.load('NeroProducer.Skim.infoProducerSequence_cff')
 process.load('NeroProducer.Nero.Nero_cfi')
+process.load('NeroProducer.Nero.NeroMonotop_cfi')
```
Comment (amarini) [inline Nero/test/testNero.py]: this should not be in the default.

File: Nero/test/testNero.py [hunk 2]
```diff
@@ -40,7 +40,7 @@
 # the size of the output by prescaling the report of the event number
 process.MessageLogger.cerr.FwkReport.reportEvery = 1000
 
-process.maxEvents = cms.untracked.PSet( input = cms.untracked.int32(-1) )
+process.maxEvents = cms.untracked.PSet( input = cms.untracked.int32(1000) )
```
Comment (amarini) [inline Nero/test/testNero.py]: this should not be default.

**Comment by sidnarayanan** [comment]
Sorry, forgot about that. Fixed

**Comment by sidnarayanan** [comment]
Sorry, I think this PR should be re-opened. I found two issues:
1) I had a bug in subjet b-tagging. This is fixed
2) (unrelated to fat jets, but I continued working on this branch) We are currently getting puppi MET straight from mini AOD, which is not to be trusted. I am working on recalculating type 1-corrected puppi MET.

I should be ready to push these commits tonight or tomorrow morning at the latest.

**Comment by yiiyama** [comment]
OK I clicked Revert without knowing the consequence; it seems that it created a new branch named revert-159-fatjet-updates. Probably intended to be pulled to your local and be applied the fixes.

**Comment by amarini** [comment]
I think that revert just create a branch to UNDO the PR. You need to have a branch that include the PR and the fixes on top.

**Comment by yiiyama** [comment]
OK, so then it's better to create a new branch from the current master and make another PR.

**Comment by amarini** [comment]
you can also continue to commit here (in the branch linked to this PR) and merge it again.
As you can see as long as there are no commit there will be nothing to merge.

**Pull request merged**